### DISCUSSION TITLE
Custom scripts are gone

### DISCUSSION THREAD

**Discussion by nomad-cyanide** [netbox-community/netbox#17557] [Help Wanted!] [answered]
After I upgraded from 3.6.9 to 4.0.11 I noticed that a few of the Reports were failing, so I set about fixing them, but no matter what I did, the changes were not being read in Netbox for GUI execution.
I then tried simply deleting the script-files from disc, to clean up. Restarted Netbox services. 
Scripts was still visible in GUI.... ?
I then tried deleting them in the GUI as well.
After that I put the scripts back in the scripts-folder and restarted Netbox services.
No scripts visible in GUI at all... ??
I even tried running an upgrade.sh, no change.

How do I get Netbox GUI to show the scripts that are on the disc?

A little about the setup:
Basically default setup regarding Netbox, with this addition:
I changed the scripts root in config away from default to: SCRIPTS_ROOT = '/opt/NB_customizations/scripts'
a git repo containing the scripts are then cloned into /opt/NB_customizations/scripts
This setup worked in 3.6.9.

**Comment by candlerb** [comment, marked as answer]
> How do I get Netbox GUI to show the scripts that are on the disc?

For standalone scripts in SCRIPTS_ROOT, you have to import them through the GUI. Dropping them into the filesystem by itself is no longer sufficient. Remove the file from the filesystem, then use Scripts > Add, Upload file. 

However if you want to keep your scripts in a git repository then I believe the official way now is to add a "Data source" to Netbox for the git repo, and then you can create scripts in the GUI which reference a file in that data source. I've never tried this myself though.

**Reply by nomad-cyanide** [reply]
> For standalone scripts in SCRIPTS_ROOT, you have to import them through the GUI. Dropping them into the filesystem by itself is no longer sufficient. Remove the file from the filesystem, then use Scripts > Add, Upload file.

Ok, didn't know that. Thank you.
 
> However if you want to keep your scripts in a git repository then I believe the official way now is to add a "Data source" to Netbox for the git repo, and then you can create scripts in the GUI which reference a file in that data source. I've never tried this myself though.

I'll test if it works with our internal repo-server, but if not, perhaps using a Data source of type: Local directory for /opt/NB_customizations/scripts would work as well.
I'll be a testing...

**Reply by nomad-cyanide** [reply]
> > However if you want to keep your scripts in a git repository then I believe the official way now is to add a "Data source" to Netbox for the git repo, and then you can create scripts in the GUI which reference a file in that data source. I've never tried this myself though.
> 
> I'll test if it works with our internal repo-server, but if not, perhaps using a Data source of type: Local directory for /opt/NB_customizations/scripts would work as well. I'll be a testing...

I tested and the Git option is straight out the window, because the username AND password is clearly visible in the web-interface. My CISO would go ballistic if I did something like that, but it works with cloning the repo to local harddrive, like /opt/scripts and creating a datasource of local type pointing to that. After a sync, scripts can be added and with the Auto-sync enabled, I just need to do a git pull, to update the scripts locally on Netbox.

**Reply by candlerb** [reply]
> I tested and the Git option is straight out the window

Could you use ssh public/private key to authenticate to the git server? I haven't tried git data sources myself, but I've read others do it. #14941, https://github.com/netbox-community/netbox/issues/15394#issuecomment-2016409907, https://github.com/netbox-community/netbox/issues/15394#issuecomment-2125648767
